File: algorithms/java/CrossSort.java
public class CrossSort
{
	public static void sort(int[] array)
	{
		int j = 0;
		int left = 0;
		int right = array.length - 1;
		boolean sorted = false;

		while(sorted == false)
		{
			sorted = true;

			for(int i = left; i < right; i++)
			{
				j = (array.length - i) - 1;

				if(i < j && array[i] > array[j])
				{
					sorted = false;
					int a = array[i];
					int b = array[j];
					array[i] = b;
					array[j] = a;
				}

				if(array[i] > array[i + 1])
				{
					sorted = false;
					int a = array[i];
					int b = array[i + 1];
					array[i] = b;
					array[i + 1] = a;
				}

				if(array[j] < array[j - 1])
				{
					sorted = false;
					int a = array[j];
					int b = array[j - 1];
					array[j] = b;
					array[j - 1] = a;
				}
			}

			left++;
			right--;
		}
	}
}

File: algorithms/go/cross_sort.go
package algorithms

func CrossSort(slice []int) {
	if(len(slice) < 2) {
		return
	}

	l := 0
	r := 0
	left := 0
	right := len(slice) - 1

	for left < right {
		if slice[left] > slice[right] {
			temp1 := slice[left]
			temp2 := slice[right]

			slice[left] = temp2
			slice[right] = temp1
		}

		l = left
		r = right

		for l < right {
			if slice[l] > slice[l + 1] {
				temp1 := slice[l]
				temp2 := slice[l + 1]

				slice[l] = temp2
				slice[l + 1] = temp1
			}

			if slice[r] < slice[r - 1] {
				temp1 := slice[r]
				temp2 := slice[r - 1]

				slice[r] = temp2
				slice[r - 1] = temp1
			}

			l++
			r--
		}

		left++
		right--
	}
}